src/main/java/org/dataloader/nextgen/Dispatcher.java

```diff
@@ -0,0 +1,297 @@
+     */
+    public Dispatcher (Executor executor) {
+        this(executor, Command::executeWith);
+    }
+    
+    /**
+     * Creates a new instance with specified ForkJoinPool.
+     * 
+     * @param executor executor to run Dispatcher tasks
+     */
+    public Dispatcher (ForkJoinPool executor) {
+        this(executor, Command::forkWith);
+    }
+
+    /**
+     * Creates a new instance that uses ForkJoinPool.commonPool()
+     */
+    public Dispatcher () {
+        this(ForkJoinPool.commonPool());
+    }
+    
+    private Dispatcher (Executor executor, BiFunction<Command, Executor, Command> invoker) {
+        Objects.requireNonNull(executor);
+        Objects.requireNonNull(invoker);
+        
+        this.executor = executor;
+        this.invoker = invoker;
+    }
+
+    @Override
+    protected void finalize() throws Throwable {
+        close();
+    }
+
+    /**
+     * Registers an new AutoDataLoader with this Dispatcher.
+     * Dispatcher will associate current thread with the new data loader
+     * and it will monitor the state in order to trigger data loader dispatch
+     * 
+     * @param dataLoader a data loader to monitor
+     * @return this instance to allow method chaining
+     */
+    public synchronized Dispatcher register (AutoDataLoader dataLoader) {
+        Objects.requireNonNull(dataLoader);
+        
+        dataLoaders.put(dataLoader, Thread.currentThread());
+        return this;
+    }
+    
+    /**
+     * Unregisters a AutoDataLoader from this Dispatcher.
+     * 
+     * @param dataLoader a data loader to unregister
+     * @return this instance to allow method chaining
+     */
+    public synchronized Dispatcher unregister (AutoDataLoader dataLoader) {
+        Objects.requireNonNull(dataLoader);
+        
+        dataLoaders.remove(dataLoader);
+        return this;
+    }
+    
+    private synchronized Thread ownerOf (AutoDataLoader dataLoader) {
+        return dataLoaders.get(dataLoader);
+    }
+    
+    /**
+     * Requests to schedule batching for the specified data loader
+     * @param dataLoader data loader to schedule dispatch
+     */
+    public void scheduleBatch (AutoDataLoader dataLoader) {
+        LOGGER.debug("dispatch requested for {}", dataLoader);
+        request(new DispatchCommand(dataLoader));
+    }
+    
+    protected void request (Command command) {
+        Objects.requireNonNull(command);
+        
+        if (running.compareAndSet(false, true)) {
+            LOGGER.debug("scheduling execution for {}", command);
+            invoker.apply(command, executor);
+        } else {
+            LOGGER.debug("enqued command {}", command);
+            commands.offer(command);
+        }
+    } 
+            
+    private static boolean isWaiting (Thread thread) {
+        switch (thread.getState()) {
+            case WAITING:
+            case TIMED_WAITING:
+                return true;
+        }
+        
+        return false;
```

Comment:
how does `java.lang.Thread.State#BLOCKED` apply here? is that not a waiting state?

Reply:
> BLOCKED
A thread that is blocked waiting for a monitor lock is in this state.

As far as I understand a Thread enters `BLOCKED` state when it tries to acquire a lock entering a critical section (like `synchronized` block to `Lock.lock()`). This state is different from `WAITING` or `TIMED_WAITING` state when a thread waits to be notified. When a thread waits for CF to complete, it has one of those states, not the `BLOCKED` one.